include/internal/event_queue.h

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_INTERNAL_EVENT_QUEUE_H
+# define OSSL_INTERNAL_EVENT_QUEUE_H
+# pragma once
+
+# include "internal/priority_queue.h"
+# include "internal/time.h"
+
+/*
+ * Opaque type holding an event.
+ */
+typedef struct ossl_event_st OSSL_EVENT;
+
+DEFINE_PRIORITY_QUEUE_OF(OSSL_EVENT);
+
+/*
+ * Public type representing an event queue, the underlying structure being
+ * opaque.
+ */
+typedef struct ossl_event_queue_st OSSL_EVENT_QUEUE;
+
+/*
+ * Generic callback function type.
+ */
+typedef int ossl_event_callback_fn(OSSL_EVENT_QUEUE *queue, uint32_t type,
+                                   void *ctx,
+                                   void *payload, size_t payload_size);
+
+/*
+ * Public type representing a event queue entry.
+ * It is (internally) public so that it can be embedded into other structures,
+ * it should otherwise be treated as opaque.
+ */
+struct ossl_event_st {
+    uint32_t type;              /* What type of event this is               */
+    uint32_t priority;          /* What priority this event has             */
+    OSSL_TIME when;             /* When the event is scheduled to happen    */
```

Comment:
I do think we can combine type and priority, and just call it type.

Reply:
It's tempting to remove `type` altogether since it's not something the event queue actually cares about anymore.